async-nats/src/jetstream/pull.rs

```diff
@@ -0,0 +1,192 @@
+        .await
+    }
+
+    pub async fn process(&mut self, batch: usize) -> PullStream<'_> {
+        PullStream::new(self, batch)
+    }
+}
+
+pub struct BatchStream<'a> {
+    pending_messages: usize,
+    subscriber: &'a mut Subscriber,
+    context: Context,
+}
+
+impl<'a> BatchStream<'a> {
+    async fn batch(
+        batch: BatchOptions,
+        consumer: &'a mut Consumer<PullConsumerConfig>,
+    ) -> Result<BatchStream<'a>, Error> {
+        if consumer.subscription.is_none() {
+            consumer.subscription = Some(
+                consumer
+                    .context
+                    .client
+                    .subscribe(consumer.inbox.clone())
+                    .await?,
+            );
+        }
+        consumer.request_batch(batch).await?;
+
+        Ok(BatchStream {
+            pending_messages: batch.batch,
+            subscriber: consumer.subscription.as_mut().unwrap(),
+            context: consumer.context.clone(),
+        })
+    }
+}
+
+impl<'a> Stream for BatchStream<'a> {
+    type Item = Result<JetStreamMessage, Error>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        if self.pending_messages == 0 {
+            return std::task::Poll::Ready(None);
+        }
+        match self.subscriber.receiver.poll_recv(cx) {
+            std::task::Poll::Ready(message) => {
+                if let Some(message) = message {
+                    if message.status == Some(StatusCode::TIMEOUT) {
+                        return Poll::Ready(None);
+                    }
+                    if message.status == Some(StatusCode::IDLE_HEARBEAT) {
+                        return std::task::Poll::Pending;
+                    }
+                    self.pending_messages -= 1;
+                    Poll::Ready(Some(Ok(JetStreamMessage {
+                        context: self.context.clone(),
+                        message,
+                    })))
+                } else {
+                    Poll::Ready(None)
+                }
+            }
+            std::task::Poll::Pending => std::task::Poll::Pending,
+        }
+    }
+}
+
+pub struct PullStream<'a> {
+    consumer: &'a mut Consumer<PullConsumerConfig>,
+    batch: usize,
+}
+
+impl<'a> PullStream<'a> {
+    fn new(consumer: &'a mut Consumer<PullConsumerConfig>, batch: usize) -> PullStream<'a> {
+        PullStream { consumer, batch }
+    }
+
+    pub fn process(&'a mut self) -> impl Stream<Item = Result<JetStreamMessage, Error>> + 'a {
+        try_stream! {
+            loop {
+                let  stream = self.consumer.batch(self.batch, None).await?;
+                for await message in stream {
+                    yield message?;
+                }
+            }
+        }
+    }
+}
+
+/// Used for next Pull Request for Pull Consumer
+#[derive(Debug, Default, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
+pub struct BatchOptions {
```

Comment:
Would like to not have this options bag, and instead expose it as different overloads/iterators. Always nicer t call flat `consumer.batch_<variant>()` calls.

Reply:
Keep in mind we have different behaviors depending on passed values:

* expires > 0 + no wait: true -> waits for first message (even if not available) then returns No Messages as soon as there are no no messages for a Consumer
* expires : 0 + no wait: true -> does not wait for first messages. Returns No Messages whenever there are no messages
* expires: whatever, no_wait: false -> waits until expires or fulfills batch.

Add to that max_bytes and heartbeat ... and we would end up with A LOT of variants. To many to have ergonomic API. Keep in mind this is not only an option bag, but an actual serialized struct as a json going via the API that we don't want to expose.

I would instead create low-level API that exposes those behaviors. But that is A LOT of overloads if we don't have arguments passed as `Option` or config bag or builder.

Reply:
> Keep in mind we have different behaviors depending on passed values:

Sounds like an argument FOR named variants that give us good documentation blocks? no waits sounds like a try variant for instance.

Reply:
We already have two variants - fetch and process (one with no wait true, on with false.).
We're missing just the third variant and you just found a nice name for it.

We could actually hide `BatchConfig` from the user, but we said that the low level API should be enabled for power users.

The only question is how to enable passing hearbet/max_bytes/fetch to those 3 methods.

Reply:
I would have first pass with current approach and extend while writing docs as a follow up PR.
@caspervonb WDYT?